backend/FwHeadless/Controllers/MediaFileController.cs

```diff
@@ -0,0 +1,385 @@
+        }
+        catch (NotFoundException) { return TypedResults.NotFound(); }
+        catch (ProjectIdRequiredForNewFiles) { return TypedResults.BadRequest(FileUploadErrorMessage.ProjectIdRequiredForNewFiles); }
+        catch (UploadedFilesCannotBeMovedToNewProjects) { return TypedResults.BadRequest(FileUploadErrorMessage.UploadedFilesCannotBeMovedToNewProjects); }
+        catch (ProjectFolderNotFoundInFwHeadless) { return TypedResults.BadRequest(FileUploadErrorMessage.ProjectFolderNotFoundInFwHeadless); }
+        catch (FileTooLarge)
+        {
+            var detail = $"Max upload size is {maxUploadSize.ToString(NumberFormatInfo.InvariantInfo)} bytes, try reducing image quality or downsampling audio";
+            return TypedResults.Problem(statusCode: 413, detail: detail);
+        }
+        filename = Path.GetFileName(mediaFile.Filename); // Strip GUID prefix for consistency with create-file path; we'll be adding it back later
+        var project = await lexBoxDb.Projects.FindAsync(projectId);
+        if (project is null) throw new NotFoundException();
+        var projectFolder = config.Value.GetProjectFolder(project.Code, projectId.Value);
+        await WriteFileAndUpdateMediaFileMetadata(lexBoxDb, mediaFile, projectFolder, fileId.Value, filename, file, maxUploadSize);
+        // Add ETag to the POST results so uploaders could, in theory, save it and use it later in a GET operation
+        var entityTag = mediaFile.Metadata!.Sha256Hash;
+        httpContext.Response.Headers.ETag = $"\"{entityTag}\"";
+        var responseBody = new PostFileResult(fileId.Value);
+        if (replacingExistingFile)
+        {
+            return TypedResults.Ok(responseBody);
+        }
+        else
+        {
+            // TODO: Put "/api/media" into a constant so if it changes in MediaFileRoutes it will change here as well
+            var newLocation = $"/api/media/{fileId}";
+            return TypedResults.Created(newLocation, responseBody);
+        }
+    }
+
+    private static async Task<long> WriteFileToDisk(string filePath, Stream contents)
+    {
+        if (contents is null) return 0;
+        long startPosition = 0;
+        try
+        {
+            startPosition = contents.Position;
+        }
+        catch { }
+        // First write to temp file, then move file into place, overwriting existing file
+        // That way files will be replaced atomically, and a failure halfway through the process won't result in the existing file being lost
+        var tempFile = Path.Join(Path.GetDirectoryName(filePath), Path.GetRandomFileName());
+        await using (var writeStream = File.Open(tempFile, FileMode.CreateNew, FileAccess.Write, FileShare.ReadWrite))
+        {
+            await contents.CopyToAsync(writeStream);
+        }
+        File.Move(tempFile, filePath, overwrite: true);
```

Comment:
if this fails for some reason then the temp file will be left in place and never cleaned up.

Reply:
Done.